### DISCUSSION TITLE
Frontend unit and end-to-end tests

### DISCUSSION THREAD

**Discussion by patryk20120** [langgenius/dify#4274] [General]
Hello everyone,

I hope this message finds you well. I have been exploring the repository, and I am impressed with the robust development and active contributions from the community. As I've navigated through the project, I observed an opportunity where I believe I could contribute significantly—specifically within the area of frontend testing using Next.js.

I propose to assist by developing a suite of unit and end-to-end tests. My intention is to ensure that the frontend features and components remain stable and are not adversely affected by incremental community contributions. Implementing these tests would help verify that changes in one segment do not unintentionally disrupt other parts of the application, an essential factor for maintaining high-quality code in collaborative projects.

Furthermore, integrating these tests into your CI/CD pipeline could automate the verification process, making it more efficient and error-resistant. 

I understand the pressures and workload that the team might currently be facing, and I believe that this contribution could be beneficial. This could potentially free up valuable developer resources, allowing the team to focus more on critical tasks such as feature development and optimization.

Would you be open to discussing this proposal further? I am eager to contribute to Dify and I believe this enhancement will aid in fortifying the project’s stability.

Thank you for considering my suggestion. I look forward to the possibility of contribution.

**Comment by perzeuss** [comment]
I suggested this before. Problem is, that Dify has still not reached version 1.0 and a lot of things might change. Adding tests now could make Dify more stable but also will slow down the development time. I am a fan of high test coverage but I understand that running tests locally as well as in the CI will also affect the development speed. The team is shipping very fast right know because they don't have to worry about failing tests or waiting for test pipelines to finish.

However, I'd like to help increasing test coverage too. I personally would love to have cypress e2e tests for the frontend. I am a fan of gherkin and cucumber which allows you to not only test things but also use the feature files as documentation.

**Comment by patryk20120** [comment]
Sorry, I did my research very badly if this was suggested.

I think it might be a good point to start as the project still develops new components. Starting testing when the project is already robust will have more disadvantages.

I don't see any issues that we could start writing tests on the side making sure that most of the things will be tested. On the side I mean that the team does not need to take care about them as they develop new features to ship them faster. 

The tests might be added later on by us or by someone else. In this case tests could help developers to check how many test were passing before and after the development to have a good overview of things. Of course they do not need to be in CI/CD for now but will be huge boost later on. 

What do you think?

**Comment by crazywoola** [comment]
Hello there, thanks for your suggestions, our team are in a rapid development process, we are trying to make this product right to the market as soon as possible, we are a small group of people here. So we do not have enough time to do test stuffs, personally I would like to add more test, actually I was add some test related packages like `faker` and etc. Pascal and I had a discussion before about this, and there was s issue complaining that there are no tests at all in frontend but solid tests in backend. :) Anyway, you are welcome to do all kinds contributions here.

**Comment by crazywoola** [comment]
Hello @patryk20120 It's been a while. 
Finally we are going to add some tests. See this pr https://github.com/langgenius/dify/pull/6426
